src/finalizer/utils/helios.ts

```
@@ -0,0 +1,887 @@
+      message: `Found ${events.length} RelayedCallData events on L2 (${l2ChainId})`,
+      spokePoolAddress: l2SpokePoolAddress,
+      fromBlock: l2SearchConfig.fromBlock,
+      toBlock: l2SearchConfig.toBlock,
+    });
+
+    // Return a Set of nonces (as strings for easy comparison)
+    return new Set<string>(events.map((event) => event.args.nonce.toString()));
+  } catch (error) {
+    logger.warn({
+      at: `Finalizer#heliosL1toL2Finalizer:getL2RelayedNonces:${l2ChainId}`,
+      message: `Failed to query RelayedCallData events from L2 (${l2ChainId})`,
+      spokePoolAddress: l2SpokePoolAddress,
+      error,
+    });
+    return null; // Return null on error
+  }
+}
+
+/**
+ * --- Get Proofs for Unfinalized Messages ---
+ * Processes messages needing proof+execution by interacting with the ZK Proof API.
+ * Returns a list of successfully retrieved proofs.
+ */
+async function processUnfinalizedHeliosMessages(
+  logger: winston.Logger,
+  messagesToProcess: PendingHeliosMessage[],
+  l2SpokePoolClient: SpokePoolClient,
+  l1ChainId: number
+): Promise<SuccessfulProof[]> {
+  // Filter within the function just in case, though the caller should have already filtered
+  const unfinalizedMessages = messagesToProcess.filter((m) => m.status === "NeedsProofAndExecution");
+  if (unfinalizedMessages.length === 0) {
+    return [];
+  }
+
+  const l2ChainId = l2SpokePoolClient.chainId;
+  const l2Provider = l2SpokePoolClient.spokePool.provider;
+  const apiBaseUrl = process.env.HELIOS_PROOF_API_URL;
+
+  if (!apiBaseUrl) {
+    logger.error({
+      at: `Finalizer#heliosL1toL2Finalizer:processUnfinalizedHeliosMessages:${l2ChainId}`,
+      message: "HELIOS_PROOF_API_URL environment variable not set. Cannot process Helios messages.",
+    });
+    return [];
+  }
+
+  const hubPoolStoreInfo = CONTRACT_ADDRESSES[l1ChainId]?.hubPoolStore;
+  if (!hubPoolStoreInfo?.address) {
+    logger.error({
```

we should throw here instead of returning []